xwiki-platform-core/xwiki-platform-search/xwiki-platform-search-solr/xwiki-platform-search-solr-api/src/main/java/org/xwiki/search/solr/internal/search/V170500000SearchCoreMigration.java

```diff
@@ -0,0 +1,140 @@
+    private EntityReferenceSerializer<String> localSerializer;
+
+    @Override
+    public long getVersion()
+    {
+        return 170500000;
+    }
+
+    @Override
+    public void migrate(XWikiSolrCore core) throws SolrException
+    {
+        // Add the docId field to the schema
+        this.solrSchema.setStringField(core, FieldUtils.DOC_ID, false, false);
+
+        // Commit
+        this.solrSchema.commit(core);
+
+        // Make sure all documents have the docId field
+        try {
+            setDocId(core);
+        } catch (Exception e) {
+            throw new SolrException("Failed to add the missing docId values to the Solr search code index", e);
+        }
+    }
+
+    private void setDocId(XWikiSolrCore core) throws SolrServerException, IOException, SolrException
+    {
+        List<SolrInputDocument> solrDocuments = new ArrayList<>();
+
+        SolrQuery query = new SolrQuery("-" + FieldUtils.DOC_ID + ":* OR -" + FieldUtils.FULLNAME + ":*");
+        query.setFields(FieldUtils.SPACE_EXACT, FieldUtils.NAME_EXACT);
+        query.setRows(10000);
+
+        String cursorMark = CursorMarkParams.CURSOR_MARK_START;
+        boolean done = false;
+        while (!done) {
+            query.set(CursorMarkParams.CURSOR_MARK_PARAM, cursorMark);
+            QueryResponse response = core.getClient().query(query);
+            String nextCursorMark = response.getNextCursorMark();
+            setDocId(response, solrDocuments);
+            if (cursorMark.equals(nextCursorMark)) {
+                done = true;
+            }
+            cursorMark = nextCursorMark;
+        }
+
+        // Send docId values
+        core.getClient().add(solrDocuments);
+
+        // Commit
+        this.solrSchema.commit(core);
+    }
+
+    private void setDocId(QueryResponse response, List<SolrInputDocument> solrDocuments)
+    {
+        for (SolrDocument readDocument : response.getResults()) {
+            String name = this.solrUtils.get(FieldUtils.NAME_EXACT, readDocument);
+            String space = this.solrUtils.get(FieldUtils.SPACE_EXACT, readDocument);
+            if (name != null && space != null) {
+                SpaceReference spaceReference = this.spaceResolver.resolve(space);
+                DocumentReference documentReference = new DocumentReference(name, spaceReference);
+                String fullName = this.localSerializer.serialize(documentReference);
+                long docId = new XWikiDocument(documentReference).getId();
+
+                SolrInputDocument writeDocument = new SolrInputDocument();
+
+                this.solrUtils.setId(this.solrUtils.getId(readDocument), writeDocument);
+
+                this.solrUtils.setAtomic(SolrUtils.ATOMIC_UPDATE_MODIFIER_SET, FieldUtils.DOC_ID, docId, writeDocument);
+                this.solrUtils.setAtomic(SolrUtils.ATOMIC_UPDATE_MODIFIER_SET, FieldUtils.FULLNAME, fullName,
+                    writeDocument);
+
+                solrDocuments.add(writeDocument);
```

Comment:
Did you test that this actually works? According to the [documentation](https://solr.apache.org/guide/solr/9_4/indexing-guide/partial-document-updates.html#field-storage), this requires quite some conditions on all fields and when I read this, I doubted that our search core would fulfill this.

Reply:
Surprisingly, the only field I could find which is not stored and not docValue (thanks to your new lowercase type) is `attcontent_` (indexed attachment content for Locale.ROOT), but strangely other `attcontent_<locale>` are stored (all the dynamic `*_<locale>` are `stored="true"`, which is actually surprising since I don't really see much point in stored those).

Reply:
As we didn't force a re-index for that field, users upgrading from older versions might still have the older fields, though. And yes, this sounds very inconsistent indeed.

Reply:
I actually noticed other unrelated inconsistencies causing problems with some old Solr index while testing, so I'm indeed afraid I don't really have much choice than top ditch the atomic update solution as being too fragile...